src/server/db_slice.cc

```diff
@@ -396,8 +397,23 @@ tuple<PrimeIterator, ExpireIterator, bool> DbSlice::AddOrFind2(const Context& cn
     }
   }
 
-  PrimeEvictionPolicy evp{cntx, (bool(caching_mode_) && !owner_->IsReplica()),
-                          int64_t(memory_budget_ - key.size()), ssize_t(soft_budget_limit_), this};
+  // In case we are loading from rdb file or replicating we want to disable conservative memory
+  // checks (inside PrimeEvictionPolicy::CanGrow) and reject insertions only after we pass max
+  // memory limit. When loading a snapshot created by the same server configuration (memory and
+  // number of shards) we will create a different dash table segment directory tree, because the
+  // tree shape is related to the order of entries insertion. Therefore when loading data from
+  // snapshot or from replication the conservative memory checks might fail as the new tree might
+  // have more segments. Because we dont want to fail loading a snapshot from the same server
+  // configuration we disable this checks on loading and replication.
+  bool apply_memory_limit =
+      !owner_->IsReplica() && !(ServerState::tlocal()->gstate() == GlobalState::LOADING);
+
+  PrimeEvictionPolicy evp{cntx,
+                          (bool(caching_mode_) && !owner_->IsReplica()),
+                          int64_t(memory_budget_ - key.size()),
+                          ssize_t(soft_budget_limit_),
+                          this,
+                          apply_memory_limit};
 
   // If we are over limit in non-cache scenario, just be conservative and throw.
   if (!caching_mode_ && evp.mem_budget() < 0) {
```

Comment:
@romange Redis does not limit memory when loading rdb file, as I wrote in another PR comment. Do you think we should do the same? If I change this if to check apply_memory_limit we will not fail on memory limit when loading at all. Currently I removed only the conservative memory limit

Reply:
@romange Based on our chat, I updated the PR to not apply memory checks when replicating or loading